byterun/floats.c

```diff
@@ -261,8 +268,17 @@ static int caml_float_of_hex(const char * s, double * res)
      on several architectures. */
   f = (double) (int64_t) m;
   /* Adjust exponent to take decimal point and extra digits into account */
-  if (dec_point >= 0) exp = exp + (dec_point - n_bits);
-  exp = exp + x_bits;
+  {
+    int adj = x_bits;
+    if (dec_point >= 0) adj = adj + (dec_point - n_bits);
+    /* saturated addition exp + adj */
+    if (adj > 0 && exp > INT_MAX - adj)
+      exp = INT_MAX;
+    else if (adj < 0 && exp < INT_MIN - adj)
+      exp = INT_MIN;
+    else
+      exp = exp + adj;
+  }
```

Comment:
The overflow detection part seems correct.

Reply:
A readability+performance suggestion: you can add a `caml_iadd_overflow` function to `misc.h` and make use of it here.

Reply:
Hmm the functions there have a different semantic (modulo operation). Even with the GCC builtin there will be tests and branches so I'm not sure there's much gain in performance.

Reply:
Oh, sure: it's saturated arithmetic. Sorry. Looks good enough as it is.